src/Core/Injector/Injector.php

```diff
@@ -966,12 +966,13 @@ public function unregisterObjects($types)
      *
      * @throws NotFoundExceptionInterface  No entry was found for **this** identifier.
      *
-     * @param string $name The name of the service to retrieve. If not a registered
+     * @template T of object
+     * @param class-string<T> $name The name of the service to retrieve. If not a registered
      * service, then a class of the given name is instantiated
      * @param bool $asSingleton If set to false a new instance will be returned.
      * If true a singleton will be returned unless the spec is type=prototype'
      * @param array $constructorArgs Args to pass in to the constructor. Note: Ignored for singletons
-     * @return mixed Instance of the specified object
+     * @return T|mixed Instance of the specified object
```

Comment:
`T|mixed`?

Shouldn't this be either `T` or `mixed` but not both?

Applies to all instances of `@return T|mixed`

Reply:
`T|mixed` does mean "`T` or `mixed` but not both" (that's what union types explicitly mean).... so I'm not sure what you mean?

Reply:
I reckon use:

```php
@param class-string<T>|string $name
...
@return T|mixed
```

Which would imply that if a class-string is used you get a type of `T`, but if you use a regular string you'd get a `mixed` type.

Reply:
Either have `@return T` or `@return mixed` but not `@return T|mixed`

`mixed` means "everything" (even null), so T is a subset of `mixed`.  This means there's no point having `T|mixed` since it's essentially just `mixed`.

However I'm not sure why we have `mixed` here since this should just be returning `T` the whole time right unless we throw an exception?

Reply:
Note that quite often you'll pass FQCN in so the generic typehint is useful to add - and when you _don't_ pass a FQCN you'll need to add a `@var` comment, the same as before this change was made. So this is a net positive change.

Reply:
I've made the change Cam suggested.